icu4c/source/common/rbbi_cache.cpp

```diff
@@ -420,50 +426,141 @@ UBool RuleBasedBreakIterator::BreakCache::populateNear(int32_t position, UErrorC
     return true;
 }
 
+static inline bool isNumber(UChar32 ch) {
+    UErrorCode status = U_ZERO_ERROR;
+    static UnicodeSet numberSet;
+    if(numberSet.isEmpty()) {
+        numberSet.applyPattern(UnicodeString(u"[:Nd:]"), status);
+        if (U_SUCCESS(status)) {
+            numberSet.compact();
+        }
+    }
+    return numberSet.contains(ch);
+}
+
+static inline bool isCj(UChar32 ch) {
+    UErrorCode status = U_ZERO_ERROR;
+    static UnicodeSet cjSet;
```

Comment:
I don't think this is thread safe. you have to use  static icu::UInitOnce with umtx_initOnce to create the UnicodeSet. Otherwise it will cause race condiction to break TSAN under multiple thread env.
Take a look at how is "static UnicodeSet *uni32Singleton" to created and used in common/uniset_props.cpp
In there the code also create a UnicodeSet, very similar to your use case. do the same thing here. 
you need to have a static UnicodeSet * and have a static icu::UInitOnce that = to  U_INITONCE_INITIALIZER
you need to have a static UBool U_CALLCONV  function to clean up your UnicodeSet when the library is unloaded in a dynamic loading env.

and look at 
```
// Cache some sets for other services -------------------------------------- ***
void U_CALLCONV createUni32Set(UErrorCode &errorCode) {
    U_ASSERT(uni32Singleton == NULL);
    uni32Singleton = new UnicodeSet(UNICODE_STRING_SIMPLE("[:age=3.2:]"), errorCode);
    if(uni32Singleton==NULL) {
        errorCode=U_MEMORY_ALLOCATION_ERROR;
    } else {
        uni32Singleton->freeze();
    }
    ucln_common_registerCleanup(UCLN_COMMON_USET, uset_cleanup);
}


U_CFUNC UnicodeSet *
uniset_getUnicode32Instance(UErrorCode &errorCode) {
    umtx_initOnce(uni32InitOnce, &createUni32Set, errorCode);
    return uni32Singleton;
}
```

Reply:
Yes, what Frank said.
ICU code cannot have static instances of objects with constructors or destructors. Statics must be pointers or simple primitive types.

Reply:
Thanks. Resolved